schemas/config/3.0.schema.json

```diff
@@ -1,2600 +1,2869 @@
+							},
+							{
+								"required": [
+									"cert",
+									"acme"
+								]
+							},
+							{
+								"required": [
+									"root_ca",
+									"acme"
+								]
+							},
+							{
+								"required": [
+									"client_auth_type",
+									"acme"
+								]
+							}
+						]
+					},
+					"anyOf": [
+						{
+							"required": [
+								"acme"
+							]
+						},
+						{
+							"required": [
+								"key",
+								"cert"
+							]
+						}
+					],
+					"properties": {
+						"address": {
+							"description": "Host address (or port) to bind to.",
+							"$ref": "#/definitions/HostAndPort",
+							"examples": [
+								"127.0.0.1:443",
+								":8443"
+							]
+						},
+						"acme": {
+							"description": "ACME certificates provider (Let's encrypt). Do not specify this parameter if you use `key` and `cert`.",
+							"type": "object",
+							"additionalProperties": false,
+							"properties": {
+								"certs_dir": {
+									"description": "Directory to use for certificates, private keys, Let's Encrypt configuration etc.",
+									"type": "string",
+									"default": "rr_cache"
+								},
+								"email": {
+									"description": "User email used to create a Let's Encrypt account. This is required.",
+									"type": "string",
+									"examples": [
+										"user@domain.com"
+									]
+								},
+								"alt_http_port": {
+									"description": "Alternate port for the HTTP challenge. Challenge traffic should be redirected to this port if overridden. See https://letsencrypt.org/docs/challenge-types/#http-01-challenge",
+									"type": "integer",
+									"default": 80
+								},
+								"alt_tlsalpn_port": {
+									"description": "Alternate port for the TLS-ALPN-01 challenge. Challenge traffic should be redirected to this port if overridden. See https://letsencrypt.org/docs/challenge-types/#tls-alpn-01",
+									"type": "integer",
+									"default": 443
+								},
+								"challenge_type": {
+									"type": "string",
+									"enum": [
+										"http-01",
+										"tlsalpn-01"
+									],
+									"description": "Challenge types",
+									"default": "http-01"
+								},
+								"use_production_endpoint": {
+									"description": "Whether to use the production endpoint. We recommend you use the staging endpoint to make sure everything works correctly before you deploy your certificate.",
+									"type": "boolean",
+									"default": false
+								},
+								"domains": {
+									"type": "array",
+									"minItems": 1,
+									"items": {
+										"type": "string",
+										"examples": [
+											"example.com"
+										]
+									},
+									"description": "List of domains to obtain certificates for. At least one domain is required."
+								}
+							},
+							"required": [
+								"domains",
+								"email"
+							]
+						},
+						"redirect": {
+							"description": "Whether to automatically redirect from HTTP to HTTPS.",
```

Comment:
Turned all these into reusable definitions because they're used in multiple places.